### PR TITLE
execute launching app and waiting results in parallel

### LINKED ISSUE #38: `npm run test-windows` doesn't succeed locally
When I run `npm run test-windows` the project is created, built and started successfully - but after a super short existence the app is closed again and the process hangs on this:

```
...
Deploying windows10 package to device:                                                                                                                                                    
C:\Users\Jan\AppData\Local\Temp\tmp-13588d52qYaYI9oDu\platforms\windows\AppPackages\CordovaApp.Windows10_1.0.0.0_anycpu_debug_Test\CordovaApp.Windows10_1.0.0.0_anycpu_debug.appx         
Attempting to uninstall previous application version...                                                                                                                                   
Attempting to install application...                                                                                                                                                      
Found package: C:\Users\Jan\AppData\Local\Temp\tmp-13588d52qYaYI9oDu\platforms\windows\AppPackages\CordovaApp.Windows10_1.0.0.0_anycpu_debug_Test\CordovaApp.Windows10_1.0.0.0_anycpu_debu
g.appx                                                                                                                                                                                    
                                                                                                                                                                                          
Installing app...                                                                                                                                                                         
                                                                                                                                                                                          
Success: Your app was successfully installed.                                                                                                                                             
Starting application...                                                                                                                                                                   
ActivateApplication:  io.cordova.hellocordova_h35559jr9hy9m!io.cordova.hellocordova                                                                                                       
Setting debug mode for application: io.cordova.hellocordova                                                                                                                               
-2147467262                                                                                                                                                                               
cordova-paramedic: waiting for test results                                                                                                                                               
```

then after a long wait fails with:

```
Completed tests at 9:48:32 PM
Collecting logs for the devices.
Running Command: C:\Users\Jan\AppData\Local\Temp\tmp-13588d52qYaYI9oDu\platforms\windows\cordova\log.bat --dump --mins 17
Uninstalling the app.
local-server: killing local file transfer server if it's up...
Error: waitForConnection: Seems like device not connected to local server in 300 secs
    at null._onTimeout (C:\Projects\Cordova\cordova-paramedic\lib\paramedic.js:547:24)
    at Timer.listOnTimeout (timers.js:92:15)

npm verb unsafe-perm in lifecycle true
npm info cordova-paramedic@0.5.0 Failed to exec test-windows script
npm verb stack Error: cordova-paramedic@0.5.0 test-windows: `node main.js --platform windows --plugin ./spec/testable-plugin/ --verbose`
npm verb stack Exit status 1
npm verb stack     at EventEmitter.<anonymous> (C:\nvm\v4.2.0\node_modules\npm\lib\utils\lifecycle.js:214:16)
npm verb stack     at emitTwo (events.js:87:13)
npm verb stack     at EventEmitter.emit (events.js:172:7)
npm verb stack     at ChildProcess.<anonymous> (C:\nvm\v4.2.0\node_modules\npm\lib\utils\spawn.js:24:14)
npm verb stack     at emitTwo (events.js:87:13)
npm verb stack     at ChildProcess.emit (events.js:172:7)
npm verb stack     at maybeClose (internal/child_process.js:818:16)
npm verb stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:211:5)
npm verb pkgid cordova-paramedic@0.5.0
npm verb cwd C:\Projects\Cordova\cordova-paramedic
npm ERR! Windows_NT 10.0.17134
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "test-windows" "--verbose"
npm ERR! node v4.2.0
npm ERR! npm  v2.14.7
npm ERR! code ELIFECYCLE
npm ERR! cordova-paramedic@0.5.0 test-windows: `node main.js --platform windows --plugin ./spec/testable-plugin/ --verbose`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the cordova-paramedic@0.5.0 test-windows script 'node main.js --platform windows --plugin ./spec/testable-plugin/ --verbose'.
npm ERR! This is most likely a problem with the cordova-paramedic package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node main.js --platform windows --plugin ./spec/testable-plugin/ --verbose
npm ERR! You can get their info via:
npm ERR!     npm owner ls cordova-paramedic
npm ERR! There is likely additional logging output above.
npm verb exit [ 1, true ]

npm ERR! Please include the following file with any support request:
npm ERR!     C:\Projects\Cordova\cordova-paramedic\npm-debug.log

C:\Projects\Cordova\cordova-paramedic (janpio-replace_wp8_with_browser -> origin) (cordova-paramedic@0.5.0)
```

Restarting the app "HelloCordova" also closes the app without further interaction or feedback.

### PR DESCRIPTION
**Pull request by knight9999**
This PR is basically same as the previous PR:  https://github.com/apache/cordova-paramedic/pull/51#issuecomment-487122818

This PR is based on the current master branch.

### Platforms affected
All platforms (tests only)
Main target is Windows 10.

### Motivation and Context
Fix #38 .

Without this PR, in runLocalTests following tasks are called in order.

 1.   launching test application
 2.  waiting application test results

However sometimes jasmine's tests in application are finished before the corodva-paramedic starts to wait the test results. I met this issue in Windows 10. (Please see #38)

### Description

With this PR, the above two tasks are called in parallel.
Then the cordova-paramedic can detect jasmine tests' results normally.

### Testing
In my local PC (Windows 10),

```
npm run test-windows
```

(with Windows Loopback Exemption Manager as mentioned https://github.com/apache/cordova-paramedic#windows-quirks )

### Checklist

- [ ] I've run the tests to see all new and existing tests pass
- [ ] I added automated test coverage as appropriate for this change
- [ ] Commit is prefixed with `(platform)` if this change only applies to one platform (e.g. `(android)`)
- [ ] If this Pull Request resolves an issue, I linked to the issue in the text above (and used the correct [keyword to close issues using keywords](https://help.github.com/articles/closing-issues-using-keywords/))
- [ ] I've updated the documentation if necessary

### CODE REVIEW

File: lib/paramedic.js [hunk 1]
```diff
@@ -216,37 +216,47 @@ class ParamedicRunner {
             .then(() => this.getCommandForStartingTests())
             .then((command) => {
                 this.setPermissions();
-                logger.normal('cordova-paramedic: running command ' + command);
+                const self = this;
```
Comment (erisu) [inline lib/paramedic.js]:
```suggestion
```

File: lib/paramedic.js [hunk 2]
```diff
@@ -216,37 +216,47 @@ class ParamedicRunner {
...
-                });
-            })
-            .then(() => {
-            // skipping here and not at the beginning because we need to
-            // start up the Android emulator for Appium tests to run on
-                if (!this.config.runMainTests()) {
-                    logger.normal('Skipping main tests...');
-                    return utilities.TEST_PASSED;
-                }
+                        if (self.config.getPlatformId() !== utilities.BROWSER) {
```
Comment (erisu) [inline lib/paramedic.js]:
```suggestion
                        if (this.config.getPlatformId() !== utilities.BROWSER) {
```

File: lib/paramedic.js [hunk 3]
```diff
@@ -216,37 +216,47 @@ class ParamedicRunner {
...
-                        // resolve if got results
-                        this.waitForTests().then(resolve);
-                    });
-                }
+                        // a precaution not to try to kill some other process
+                        runProcess = cp.exec(command, () => {
+                            runProcess = null;
+                        });
+                    }),
+                    Q().then(function () {
```
Comment (erisu) [inline lib/paramedic.js]:
```suggestion
                    Q().then(() => {
```

File: lib/paramedic.js [hunk 4]
```diff
@@ -216,37 +216,47 @@ class ParamedicRunner {
...
+                        // skipping here and not at the beginning because we need to
+                        // start up the Android emulator for Appium tests to run on
+                        if (!self.config.runMainTests()) {
```
Comment (erisu) [inline lib/paramedic.js]:
```suggestion
                        if (!this.config.runMainTests()) {
```

File: lib/paramedic.js [hunk 5]
```diff
@@ -216,37 +216,47 @@ class ParamedicRunner {
...
+                            logger.normal('Skipping main tests...');
+                            return utilities.TEST_PASSED;
+                        }
+
+                        // skip tests if it was just build
+                        if (self.shouldWaitForTestResult()) {
```
Comment (erisu) [inline lib/paramedic.js]:
```suggestion
                        if (this.shouldWaitForTestResult()) {
```

File: lib/paramedic.js [hunk 6]
```diff
@@ -216,37 +216,47 @@ class ParamedicRunner {
...
+                            return Q.promise((resolve, reject) => {
+                            // reject if timed out
+                                self.waitForConnection().catch(reject);
```
Comment (erisu) [inline lib/paramedic.js]:
```suggestion
                                this.waitForConnection().catch(reject);
```

File: lib/paramedic.js [hunk 7]
```diff
@@ -216,37 +216,47 @@ class ParamedicRunner {
...
+                                // resolve if got results
+                                self.waitForTests().then(resolve);
```
Comment (erisu) [inline lib/paramedic.js]:
```suggestion
                                this.waitForTests().then(resolve);
```

File: lib/paramedic.js [hunk 8]
```diff
@@ -216,37 +216,47 @@ class ParamedicRunner {
...
                 this.setPermissions();
-                logger.normal('cordova-paramedic: running command ' + command);
+                const self = this;
 
-                if (this.config.getPlatformId() !== utilities.BROWSER) {
-                    return execPromise(command);
-                }
-                console.log('$ ' + command);
+                return Q.all([
+                    Q().then(function () {
```
Comment (erisu) [inline lib/paramedic.js]:
```suggestion
                    Q().then(() => {
```

**Comment by knight9999** [comment]
@erisu Thanks for the suggestions. I accepted them.

**Pull request merged**